pychromecast/discovery.py

```diff
@@ -106,56 +162,252 @@ def get_value(key):
             return
         uuid = UUID(uuid)
 
-        services_for_uuid = self.services.setdefault(
-            uuid, ({name}, uuid, model_name, friendly_name)
+        cast_info = self._devices.setdefault(
+            uuid, CastInfo(set(), uuid, model_name, friendly_name, host, service.port)
         )
-        services_for_uuid[0].add(name)
-        self.services[uuid] = (
-            services_for_uuid[0],
-            services_for_uuid[1],
-            model_name,
-            friendly_name,
-            host,
-            service.port,
+        # Update stored information
+        cast_info.services.add(ServiceInfo(SERVICE_TYPE_MDNS, name))
+        self._devices[uuid] = CastInfo(
+            cast_info.services, uuid, model_name, friendly_name, host, service.port
         )
 
-        if callback:
-            callback(uuid, name)
+        callback(uuid, name)
 
 
-def start_discovery(listener, zeroconf_instance):
-    """
-    Start discovering chromecasts on the network.
+class HostStatus:
+    """Status of known host."""
 
-    This method will start discovering chromecasts on a separate thread. When
-    a chromecast is discovered, the callback will be called with the
-    discovered chromecast's zeroconf name. This is the dictionary key to find
-    the chromecast metadata in listener.services.
+    def __init__(self):
+        self.failcount = 0
 
-    This method returns the zeroconf ServiceBrowser object.
 
-    A CastListener object must be passed, and will contain information for the
-    discovered chromecasts. To stop discovery, call the stop_discovery method with
-    the ServiceBrowser object.
+HOSTLISTENER_CYCLE_TIME = 5
+HOSTLISTENER_MAX_FAIL = 5
+
+
+class HostBrowser(threading.Thread):
+    """"Repeateadly poll a set of known hosts."""
+
+    # pylint: disable=unused-argument
+    def __init__(self, cast_listener, devices):
+        super().__init__(daemon=True)
+        self._cast_listener = cast_listener
+        self._devices = devices
+        self._known_hosts = {}
+        self._next_update = time.time()
+        self._start_requested = False
+        self.stop = threading.Event()
+
+    def add_hosts(self, known_hosts):
+        """Add a list of known hosts to the set."""
+        for host in known_hosts:
+            self._known_hosts.setdefault(host, HostStatus())
```

Comment:
This could potentially create and throw away a lot of `HostStatus` objects. Why not check if key `host` is already set?

Reply:
Thanks, fixed!